File: frontend/src/schemas/providers.ts
import { z } from "zod";

export const CREDENTIAL_MODES = ["admin", "user", "both"] as const;

export type CredentialModeType = (typeof CREDENTIAL_MODES)[number];

export const userCredentialsSchema = z.object({
  client_id: z
    .string()
    .min(1, "Client ID is required")
    .max(500, "Client ID must be at most 500 characters"),
  client_secret: z
    .string()
    .max(2000, "Client Secret must be at most 2000 characters")
    .optional()
    .or(z.literal("")),
  label: z
    .string()
    .max(200, "Label must be at most 200 characters")
    .optional()
    .or(z.literal("")),
});

export type UserCredentialsFormData = z.infer<typeof userCredentialsSchema>;

export const connectApiKeySchema = z.object({
  api_key: z
    .string()
    .min(1, "API key is required")
    .max(8192, "API key must be at most 8192 characters"),
  label: z.string().max(200, "Label must be at most 200 characters").optional(),
});

export type ConnectApiKeyFormData = z.infer<typeof connectApiKeySchema>;

const optionalTelegramString = z.string().trim().min(1);
const TELEGRAM_BOT_USERNAME_PATTERN = /^@?[a-z][a-z0-9_]{1,28}bot$/i;
const TELEGRAM_BOT_USERNAME_MESSAGE =
  "Bot username must be 5-32 characters, start with a letter, use only letters, digits, or underscores, and end in 'bot'";

export const telegramLoginDataSchema = z.object({
  id: z.coerce.number().int().positive(),
  first_name: z.string().trim().min(1, "First name is required"),
  last_name: optionalTelegramString.optional(),
  username: optionalTelegramString.optional(),
  photo_url: z.string().url("Photo URL must be a valid URL").optional(),
  auth_date: z.coerce.number().int().positive(),
  hash: z
    .string()
    .trim()
    .regex(/^[a-fA-F0-9]{64}$/, "Invalid Telegram login hash"),
});

export const PROVIDER_TYPES = [
  "oauth2",
  "api_key",
  "device_code",
  "telegram_widget",
] as const;

export type ProviderType = (typeof PROVIDER_TYPES)[number];

const SLUG_PATTERN = /^[a-z0-9][a-z0-9-]*[a-z0-9]$/;

const baseProviderFields = {
  name: z
    .string()
    .min(2, "Name must be at least 2 characters")
    .max(100, "Name must be at most 100 characters"),
  slug: z
    .string()
    .min(2, "Slug must be at least 2 characters")
    .max(50, "Slug must be at most 50 characters")
    .regex(
      SLUG_PATTERN,
      "Slug must contain only lowercase letters, digits, and hyphens (no leading/trailing hyphens)",
    ),
  description: z
    .string()
    .max(500, "Description must be at most 500 characters")
    .optional()
    .or(z.literal("")),
  provider_type: z.enum(PROVIDER_TYPES),
  authorization_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  token_url: z.string().url("Must be a valid URL").optional().or(z.literal("")),
  revocation_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  default_scopes: z
    .string()
    .max(2000, "Scopes must be at most 2000 characters")
    .optional()
    .or(z.literal("")),
  credential_mode: z.enum(CREDENTIAL_MODES).optional(),
  supports_pkce: z.boolean().optional(),
  device_code_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  device_token_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  device_verification_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  hosted_callback_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  api_key_instructions: z
    .string()
    .max(2000, "Instructions must be at most 2000 characters")
    .optional()
    .or(z.literal("")),
  api_key_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  icon_url: z.string().url("Must be a valid URL").optional().or(z.literal("")),
  documentation_url: z
    .string()
    .url("Must be a valid URL")
    .optional()
    .or(z.literal("")),
  token_endpoint_auth_method: z
    .enum(["client_secret_post", "client_secret_basic"])
    .optional(),
  extra_auth_params: z.record(z.string(), z.string()).optional(),
  device_code_format: z.enum(["rfc8628", "openai"]).optional(),
  client_id_param_name: z
    .string()
    .max(100, "Param name must be at most 100 characters")
    .optional()
    .or(z.literal("")),
} as const;

export const createProviderSchema = z
  .object({
    ...baseProviderFields,
    client_id: z.string().optional().or(z.literal("")),
    client_secret: z.string().optional().or(z.literal("")),
  })
  .superRefine((data, ctx) => {
    const mode = data.credential_mode ?? "admin";
    if (data.provider_type === "oauth2") {
      if (!data.authorization_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Authorization URL is required for OAuth2 providers",
          path: ["authorization_url"],
        });
      }
      if (!data.token_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Token URL is required for OAuth2 providers",
          path: ["token_url"],
        });
      }
      if (mode === "admin") {
        if (!data.client_id) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: "Client ID is required for OAuth2 providers in admin mode",
            path: ["client_id"],
          });
        }
        if (!data.client_secret) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message:
              "Client Secret is required for OAuth2 providers in admin mode",
            path: ["client_secret"],
          });
        }
      } else if (!!data.client_id !== !!data.client_secret) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            "Admin fallback OAuth2 credentials must include both Client ID and Client Secret",
          path: [data.client_id ? "client_secret" : "client_id"],
        });
      }
    }
    if (data.provider_type === "device_code") {
      if (!data.authorization_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Authorization URL is required for device code providers",
          path: ["authorization_url"],
        });
      }
      if (!data.token_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Token URL is required for device code providers",
          path: ["token_url"],
        });
      }
      if (mode === "admin" && !data.client_id) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            "Client ID is required for device code providers in admin mode",
          path: ["client_id"],
        });
      }
      if (data.client_secret && !data.client_id) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Client ID is required when Client Secret is set",
          path: ["client_id"],
        });
      }
      if (!data.device_code_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Device Code URL is required for device code providers",
          path: ["device_code_url"],
        });
      }
      if (!data.device_token_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Device Token URL is required for device code providers",
          path: ["device_token_url"],
        });
      }
    }
    if (data.provider_type === "telegram_widget") {
      if (data.credential_mode && data.credential_mode !== "admin") {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            "Telegram widget providers only support admin credential mode",
          path: ["credential_mode"],
        });
      }
      if (!data.client_id_param_name?.trim()) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Bot username is required for Telegram widget providers",
          path: ["client_id_param_name"],
        });
      } else if (
        !TELEGRAM_BOT_USERNAME_PATTERN.test(data.client_id_param_name.trim())
      ) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: TELEGRAM_BOT_USERNAME_MESSAGE,
          path: ["client_id_param_name"],
        });
      }
      if (!data.client_secret?.trim()) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Bot token is required for Telegram widget providers",
          path: ["client_secret"],
        });
      }
    }
  });

export type CreateProviderFormData = z.infer<typeof createProviderSchema>;

export const updateProviderSchema = z
  .object({
    ...baseProviderFields,
    is_active: z.boolean().optional(),
    client_id: z.string().optional().or(z.literal("")),
    client_secret: z.string().optional().or(z.literal("")),
  })
  .superRefine((data, ctx) => {
    if (data.provider_type === "oauth2") {
      if (!data.authorization_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Authorization URL is required for OAuth2 providers",
          path: ["authorization_url"],
        });
      }
      if (!data.token_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Token URL is required for OAuth2 providers",
          path: ["token_url"],
        });
      }
    }
    if (data.provider_type === "device_code") {
      if (!data.authorization_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Authorization URL is required for device code providers",
          path: ["authorization_url"],
        });
      }
      if (!data.token_url) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Token URL is required for device code providers",
          path: ["token_url"],
        });
      }
    }
    if (
      data.provider_type === "telegram_widget" &&
      data.credential_mode &&
      data.credential_mode !== "admin"
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Telegram widget providers only support admin credential mode",
        path: ["credential_mode"],
      });
    }
    if (
      data.provider_type === "telegram_widget" &&
      data.client_id_param_name &&
      !data.client_id_param_name.trim()
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Bot username must not be blank",
        path: ["client_id_param_name"],
      });
    } else if (
      data.provider_type === "telegram_widget" &&
      data.client_id_param_name &&
      !TELEGRAM_BOT_USERNAME_PATTERN.test(data.client_id_param_name.trim())
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: TELEGRAM_BOT_USERNAME_MESSAGE,
        path: ["client_id_param_name"],
      });
    }
    if (
      data.provider_type === "telegram_widget" &&
      data.client_secret &&
      !data.client_secret.trim()
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Bot token must not be blank",
        path: ["client_secret"],
      });
    }
    // Note: device_code_url and device_token_url are optional on update (blank = keep current)
  });

export type UpdateProviderFormData = z.infer<typeof updateProviderSchema>;

File: backend/src/handlers/providers.rs
use axum::{
    Json,
    extract::{Path, State},
};
use serde::{Deserialize, Serialize};

use crate::AppState;
use crate::errors::{AppError, AppResult};
use crate::mw::auth::AuthUser;
use crate::services::{audit_service, provider_service, user_credentials_service};

use super::services_helpers::{require_admin, validate_base_url};

// --- Request / Response types ---

#[derive(Deserialize)]
pub struct CreateProviderRequest {
    pub name: String,
    pub slug: String,
    pub description: Option<String>,
    pub provider_type: String,
    pub credential_mode: Option<String>,
    // OAuth2 fields
    pub authorization_url: Option<String>,
    pub token_url: Option<String>,
    pub revocation_url: Option<String>,
    pub default_scopes: Option<Vec<String>>,
    pub client_id: Option<String>,
    pub client_secret: Option<String>,
    pub supports_pkce: Option<bool>,
    // Device code flow fields
    pub device_code_url: Option<String>,
    pub device_token_url: Option<String>,
    pub device_verification_url: Option<String>,
    pub hosted_callback_url: Option<String>,
    // API key fields
    pub api_key_instructions: Option<String>,
    pub api_key_url: Option<String>,
    // Display
    pub icon_url: Option<String>,
    pub documentation_url: Option<String>,
    // Auth method
    pub token_endpoint_auth_method: Option<String>,
    // Generic OAuth edge case fields
    pub extra_auth_params: Option<std::collections::HashMap<String, String>>,
    pub device_code_format: Option<String>,
    pub client_id_param_name: Option<String>,
}

impl std::fmt::Debug for CreateProviderRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CreateProviderRequest")
            .field("name", &self.name)
            .field("slug", &self.slug)
            .field("provider_type", &self.provider_type)
            .field("client_id", &self.client_id.as_ref().map(|_| "[REDACTED]"))
            .field(
                "client_secret",
                &self.client_secret.as_ref().map(|_| "[REDACTED]"),
            )
            .finish()
    }
}

#[derive(Deserialize)]
pub struct UpdateProviderRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub is_active: Option<bool>,
    pub authorization_url: Option<String>,
    pub token_url: Option<String>,
    pub revocation_url: Option<String>,
    pub default_scopes: Option<Vec<String>>,
    pub client_id: Option<String>,
    pub client_secret: Option<String>,
    pub supports_pkce: Option<bool>,
    pub device_code_url: Option<String>,
    pub device_token_url: Option<String>,
    pub device_verification_url: Option<String>,
    pub hosted_callback_url: Option<String>,
    pub api_key_instructions: Option<String>,
    pub api_key_url: Option<String>,
    pub icon_url: Option<String>,
    pub documentation_url: Option<String>,
    pub credential_mode: Option<String>,
    pub token_endpoint_auth_method: Option<String>,
    pub extra_auth_params: Option<std::collections::HashMap<String, String>>,
    pub device_code_format: Option<String>,
    pub client_id_param_name: Option<String>,
}

impl std::fmt::Debug for UpdateProviderRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("UpdateProviderRequest")
            .field("name", &self.name)
            .field("description", &self.description)
            .field("is_active", &self.is_active)
            .field("client_id", &self.client_id.as_ref().map(|_| "[REDACTED]"))
            .field(
                "client_secret",
                &self.client_secret.as_ref().map(|_| "[REDACTED]"),
            )
            .finish()
    }
}

/// Validate slug format: lowercase alphanumeric and hyphens only, no
/// leading/trailing/consecutive hyphens.
fn validate_slug(slug: &str) -> AppResult<()> {
    if slug.is_empty() || slug.len() > 100 {
        return Err(AppError::ValidationError(
            "slug must be between 1 and 100 characters".to_string(),
        ));
    }
    let valid = slug
        .chars()
        .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
        && !slug.starts_with('-')
        && !slug.ends_with('-')
        && !slug.contains("--");
    if !valid {
        return Err(AppError::ValidationError(
            "slug must contain only lowercase letters, digits, and hyphens (no leading/trailing/consecutive hyphens)".to_string(),
        ));
    }
    Ok(())
}

#[derive(Debug, Serialize)]
pub struct ProviderResponse {
    pub id: String,
    pub slug: String,
    pub name: String,
    pub description: Option<String>,
    pub provider_type: String,
    pub has_oauth_config: bool,
    pub default_scopes: Option<Vec<String>>,
    pub supports_pkce: bool,
    pub device_code_url: Option<String>,
    pub device_token_url: Option<String>,
    pub device_verification_url: Option<String>,
    pub hosted_callback_url: Option<String>,
    pub api_key_instructions: Option<String>,
    pub api_key_url: Option<String>,
    pub icon_url: Option<String>,
    pub documentation_url: Option<String>,
    pub is_active: bool,
    pub credential_mode: String,
    pub token_endpoint_auth_method: String,
    pub extra_auth_params: Option<std::collections::HashMap<String, String>>,
    pub device_code_format: String,
    pub client_id_param_name: Option<String>,
    pub requires_gateway_url: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize)]
pub struct ProviderListResponse {
    pub providers: Vec<ProviderResponse>,
}

#[derive(Debug, Serialize)]
pub struct DeleteProviderResponse {
    pub message: String,
}

fn provider_to_response(p: crate::models::provider_config::ProviderConfig) -> ProviderResponse {
    let has_oauth_config = provider_has_oauth_config(&p);

    ProviderResponse {
        id: p.id,
        slug: p.slug,
        name: p.name,
        description: p.description,
        provider_type: p.provider_type,
        has_oauth_config,
        default_scopes: p.default_scopes,
        supports_pkce: p.supports_pkce,
        device_code_url: p.device_code_url,
        device_token_url: p.device_token_url,
        device_verification_url: p.device_verification_url,
        hosted_callback_url: p.hosted_callback_url,
        api_key_instructions: p.api_key_instructions,
        api_key_url: p.api_key_url,
        icon_url: p.icon_url,
        documentation_url: p.documentation_url,
        is_active: p.is_active,
        credential_mode: p.credential_mode,
        token_endpoint_auth_method: p.token_endpoint_auth_method,
        extra_auth_params: p.extra_auth_params,
        device_code_format: p.device_code_format,
        client_id_param_name: p.client_id_param_name,
        requires_gateway_url: p.requires_gateway_url,
        created_at: p.created_at.to_rfc3339(),
        updated_at: p.updated_at.to_rfc3339(),
    }
}

/// Check whether a provider has enough configuration to start an OAuth flow.
///
/// - For `"user"` mode: true if the shared OAuth URLs are configured.
/// - For `"admin"`/`"both"` mode: true if OAuth URLs AND admin-level credentials are configured.
fn provider_has_oauth_config(p: &crate::models::provider_config::ProviderConfig) -> bool {
    let has_urls = match p.provider_type.as_str() {
        "oauth2" => p.authorization_url.is_some() && p.token_url.is_some(),
        "device_code" => {
            p.authorization_url.is_some()
                && p.token_url.is_some()
                && p.device_code_url.is_some()
                && p.device_token_url.is_some()
        }
        _ => return false,
    };

    if !has_urls {
        return false;
    }

    if p.credential_mode == "user" {
        return true;
    }

    user_credentials_service::provider_has_admin_oauth_credentials(p)
}

// --- Handlers ---

/// GET /api/v1/providers
pub async fn list_providers(
    State(state): State<AppState>,
    _auth_user: AuthUser,
) -> AppResult<Json<ProviderListResponse>> {
    let providers = provider_service::list_providers(&state.db).await?;

    let items: Vec<ProviderResponse> = providers.into_iter().map(provider_to_response).collect();

    Ok(Json(ProviderListResponse { providers: items }))
}

/// POST /api/v1/providers
pub async fn create_provider(
    State(state): State<AppState>,
    auth_user: AuthUser,
    Json(body): Json<CreateProviderRequest>,
) -> AppResult<Json<ProviderResponse>> {
    require_admin(&state, &auth_user).await?;

    if body.name.is_empty() || body.slug.is_empty() {
        return Err(AppError::ValidationError(
            "name and slug are required".to_string(),
        ));
    }

    if body.name.len() > 200 {
        return Err(AppError::ValidationError(
            "name exceeds maximum length of 200 characters".to_string(),
        ));
    }

    // Validate slug format
    validate_slug(&body.slug)?;

    // Validate provider_type
    let valid_types = ["oauth2", "api_key", "device_code"];
    if !valid_types.contains(&body.provider_type.as_str()) {
        return Err(AppError::ValidationError(format!(
            "provider_type must be one of: {}",
            valid_types.join(", ")
        )));
    }
    // Defense-in-depth: service layer also validates credential_mode
    let credential_mode = body.credential_mode.as_deref().unwrap_or("admin");
    let valid_credential_modes = ["admin", "user", "both"];
    if !valid_credential_modes.contains(&credential_mode) {
        return Err(AppError::ValidationError(format!(
            "credential_mode must be one of: {}",
            valid_credential_modes.join(", ")
        )));
    }

    let token_endpoint_auth_method = body
        .token_endpoint_auth_method
        .as_deref()
        .unwrap_or("client_secret_post");
    let valid_auth_methods = ["client_secret_post", "client_secret_basic"];
    if !valid_auth_methods.contains(&token_endpoint_auth_method) {
        return Err(AppError::ValidationError(format!(
            "token_endpoint_auth_method must be one of: {}",
            valid_auth_methods.join(", ")
        )));
    }

    if let Some(ref format) = body.device_code_format
        && !["rfc8628", "openai"].contains(&format.as_str())
    {
        return Err(AppError::ValidationError(
            "device_code_format must be 'rfc8628' or 'openai'".to_string(),
        ));
    }

    let user_id_str = auth_user.user_id.to_string();

    let oauth_config = if body.provider_type == "oauth2" {
        let authorization_url = body.authorization_url.as_ref().ok_or_else(|| {
            AppError::ValidationError(
                "authorization_url is required for OAuth2 providers".to_string(),
            )
        })?;
        let token_url = body.token_url.as_ref().ok_or_else(|| {
            AppError::ValidationError("token_url is required for OAuth2 providers".to_string())
        })?;

        // SSRF validation on OAuth provider URLs
        validate_base_url(authorization_url)?;
        validate_base_url(token_url)?;
        if let Some(ref url) = body.revocation_url {
            validate_base_url(url)?;
        }

        let client_id = body.client_id.clone();
        let client_secret = body.client_secret.clone();
        let has_client_id = client_id.is_some();
        let has_client_secret = client_secret.is_some();

        if credential_mode == "admin" {
            if !has_client_id {
                return Err(AppError::ValidationError(
                    "client_id is required for OAuth2 providers in admin mode".to_string(),
                ));
            }
            if !has_client_secret {
                return Err(AppError::ValidationError(
                    "client_secret is required for OAuth2 providers in admin mode".to_string(),
                ));
            }
        } else if has_client_id != has_client_secret {
            return Err(AppError::ValidationError(
                "OAuth2 admin fallback credentials must include both client_id and client_secret"
                    .to_string(),
            ));
        }

        Some(provider_service::OAuthProviderInput {
            authorization_url: authorization_url.clone(),
            token_url: token_url.clone(),
            revocation_url: body.revocation_url.clone(),
            default_scopes: body.default_scopes.clone(),
            client_id,
            client_secret,
            supports_pkce: body.supports_pkce.unwrap_or(false),
        })
    } else {
        None
    };

    let device_code_config = if body.provider_type == "device_code" {
        let authorization_url = body.authorization_url.as_ref().ok_or_else(|| {
            AppError::ValidationError(
                "authorization_url is required for device_code providers".to_string(),
            )
        })?;
        let token_url = body.token_url.as_ref().ok_or_else(|| {
            AppError::ValidationError("token_url is required for device_code providers".to_string())
        })?;
        let device_code_url = body.device_code_url.as_ref().ok_or_else(|| {
            AppError::ValidationError(
                "device_code_url is required for device_code providers".to_string(),
            )
        })?;
        let device_token_url = body.device_token_url.as_ref().ok_or_else(|| {
            AppError::ValidationError(
                "device_token_url is required for device_code providers".to_string(),
            )
        })?;

        // SSRF validation on all URLs
        validate_base_url(authorization_url)?;
        validate_base_url(token_url)?;
        validate_base_url(device_code_url)?;
        validate_base_url(device_token_url)?;
        if let Some(ref url) = body.device_verification_url {
            validate_base_url(url)?;
        }
        if let Some(ref url) = body.hosted_callback_url {
            validate_base_url(url)?;
        }

        let client_id = body.client_id.clone();
        if credential_mode == "admin" && client_id.is_none() {
            return Err(AppError::ValidationError(
                "client_id is required for device_code providers in admin mode".to_string(),
            ));
        }
        if body.client_secret.is_some() && client_id.is_none() {
            return Err(AppError::ValidationError(
                "client_id is required when setting a device_code client_secret".to_string(),
            ));
        }

        Some(provider_service::DeviceCodeProviderInput {
            authorization_url: authorization_url.clone(),
            token_url: token_url.clone(),
            device_code_url: device_code_url.clone(),
            device_token_url: device_token_url.clone(),
            device_verification_url: body.device_verification_url.clone(),
            hosted_callback_url: body.hosted_callback_url.clone(),
            default_scopes: body.default_scopes.clone(),
            client_id,
            client_secret: body.client_secret.clone(),
            supports_pkce: body.supports_pkce.unwrap_or(true),
        })
    } else {
        None
    };

    let api_key_config = if body.provider_type == "api_key" {
        Some(provider_service::ApiKeyProviderInput {
            api_key_instructions: body.api_key_instructions.clone(),
            api_key_url: body.api_key_url.clone(),
        })
    } else {
        None
    };

    let provider = provider_service::create_provider(
        &state.db,
        &state.encryption_keys,
        &body.name,
        &body.slug,
        &body.provider_type,
        credential_mode,
        token_endpoint_auth_method,
        oauth_config,
        api_key_config,
        device_code_config,
        body.description.as_deref(),
        body.icon_url.as_deref(),
        body.documentation_url.as_deref(),
        &user_id_str,
        body.extra_auth_params,
        body.device_code_format.as_deref(),
        body.client_id_param_name.as_deref(),
    )
    .await?;

    audit_service::log_async(
        state.db.clone(),
        Some(user_id_str),
        "provider_created".to_string(),
        Some(serde_json::json!({
            "provider_id": &provider.id,
            "slug": &provider.slug,
        })),
        None,
        None,
    );

    Ok(Json(provider_to_response(provider)))
}

/// GET /api/v1/providers/{provider_id}
pub async fn get_provider(
    State(state): State<AppState>,
    _auth_user: AuthUser,
    Path(provider_id): Path<String>,
) -> AppResult<Json<ProviderResponse>> {
    let provider = provider_service::get_provider(&state.db, &provider_id).await?;
    Ok(Json(provider_to_response(provider)))
}

/// PUT /api/v1/providers/{provider_id}
pub async fn update_provider(
    State(state): State<AppState>,
    auth_user: AuthUser,
    Path(provider_id): Path<String>,
    Json(body): Json<UpdateProviderRequest>,
) -> AppResult<Json<ProviderResponse>> {
    require_admin(&state, &auth_user).await?;

    // SSRF validation on URLs if provided
    if let Some(ref url) = body.authorization_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.token_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.revocation_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.device_code_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.device_token_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.device_verification_url {
        validate_base_url(url)?;
    }
    if let Some(ref url) = body.hosted_callback_url {
        validate_base_url(url)?;
    }

    let updates = provider_service::ProviderUpdateInput {
        name: body.name,
        description: body.description,
        is_active: body.is_active,
        authorization_url: body.authorization_url,
        token_url: body.token_url,
        revocation_url: body.revocation_url,
        default_scopes: body.default_scopes,
        client_id: body.client_id,
        client_secret: body.client_secret,
        supports_pkce: body.supports_pkce,
        device_code_url: body.device_code_url,
        device_token_url: body.device_token_url,
        device_verification_url: body.device_verification_url,
        hosted_callback_url: body.hosted_callback_url,
        api_key_instructions: body.api_key_instructions,
        api_key_url: body.api_key_url,
        icon_url: body.icon_url,
        documentation_url: body.documentation_url,
        credential_mode: body.credential_mode,
        token_endpoint_auth_method: body.token_endpoint_auth_method,
        extra_auth_params: body.extra_auth_params,
        device_code_format: body.device_code_format,
        client_id_param_name: body.client_id_param_name,
    };

    let updated =
        provider_service::update_provider(&state.db, &state.encryption_keys, &provider_id, updates)
            .await?;

    audit_service::log_async(
        state.db.clone(),
        Some(auth_user.user_id.to_string()),
        "provider_updated".to_string(),
        Some(serde_json::json!({ "provider_id": &provider_id })),
        None,
        None,
    );

    Ok(Json(provider_to_response(updated)))
}

/// DELETE /api/v1/providers/{provider_id}
pub async fn delete_provider(
    State(state): State<AppState>,
    auth_user: AuthUser,
    Path(provider_id): Path<String>,
) -> AppResult<Json<DeleteProviderResponse>> {
    require_admin(&state, &auth_user).await?;

    provider_service::delete_provider(&state.db, &provider_id).await?;

    audit_service::log_async(
        state.db.clone(),
        Some(auth_user.user_id.to_string()),
        "provider_deleted".to_string(),
        Some(serde_json::json!({ "provider_id": &provider_id })),
        None,
        None,
    );

    Ok(Json(DeleteProviderResponse {
        message: "Provider deactivated and user tokens revoked".to_string(),
    }))
}

#[cfg(test)]
mod tests {
    use chrono::Utc;

    use super::provider_has_oauth_config;
    use crate::models::provider_config::ProviderConfig;

    fn make_provider(provider_type: &str) -> ProviderConfig {
        ProviderConfig {
            id: "provider-1".to_string(),
            slug: "provider-1".to_string(),
            name: "Provider".to_string(),
            description: None,
            provider_type: provider_type.to_string(),
            authorization_url: Some("https://auth.example.com/authorize".to_string()),
            token_url: Some("https://auth.example.com/token".to_string()),
            revocation_url: None,
            default_scopes: None,
            client_id_encrypted: Some(vec![1, 2, 3]),
            client_secret_encrypted: Some(vec![4, 5, 6]),
            supports_pkce: true,
            device_code_url: Some("https://auth.example.com/device".to_string()),
            device_token_url: Some("https://auth.example.com/device/token".to_string()),
            device_verification_url: None,
            hosted_callback_url: None,
            api_key_instructions: None,
            api_key_url: None,
            icon_url: None,
            documentation_url: None,
            is_active: true,
            credential_mode: "admin".to_string(),
            token_endpoint_auth_method: "client_secret_post".to_string(),
            extra_auth_params: None,
            device_code_format: "rfc8628".to_string(),
            client_id_param_name: None,
            requires_gateway_url: false,
            created_by: "system".to_string(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }

    #[test]
    fn oauth2_requires_a_client_secret_to_be_connectable() {
        let mut provider = make_provider("oauth2");
        provider.client_secret_encrypted = None;

        assert!(!provider_has_oauth_config(&provider));
    }

    #[test]
    fn device_code_provider_can_be_configured_without_a_client_secret() {
        let mut provider = make_provider("device_code");
        provider.client_secret_encrypted = None;

        assert!(provider_has_oauth_config(&provider));
    }

    #[test]
    fn both_mode_requires_admin_credentials_for_shared_connectability() {
        let mut provider = make_provider("oauth2");
        provider.credential_mode = "both".to_string();
        provider.client_id_encrypted = None;
        provider.client_secret_encrypted = None;

        assert!(!provider_has_oauth_config(&provider));
    }

    #[test]
    fn user_mode_only_needs_oauth_urls() {
        let mut provider = make_provider("oauth2");
        provider.credential_mode = "user".to_string();
        provider.client_id_encrypted = None;
        provider.client_secret_encrypted = None;

        assert!(provider_has_oauth_config(&provider));
    }
}
